File: src/lib/client-ip.ts
const CLIENT_IP_HEADER_CANDIDATES = [
  "cf-connecting-ip",
  "true-client-ip",
  "x-forwarded-for",
  "x-real-ip"
] as const;

function normalizeClientIpCandidate(value: string | null | undefined) {
  if (typeof value !== "string") return null;
  const candidate = value.split(",")[0]?.trim();
  if (!candidate) return null;
  return candidate.slice(0, 64);
}

export function resolveForwardedClientIp(headers: Headers) {
  for (const headerName of CLIENT_IP_HEADER_CANDIDATES) {
    const candidate = normalizeClientIpCandidate(headers.get(headerName));
    if (candidate) return candidate;
  }
  return null;
}

export function copyForwardedClientIpHeaders(source: Headers, target: Headers) {
  const cfConnectingIp = normalizeClientIpCandidate(source.get("cf-connecting-ip"));
  if (cfConnectingIp) target.set("cf-connecting-ip", cfConnectingIp);

  const trueClientIp = normalizeClientIpCandidate(source.get("true-client-ip"));
  if (trueClientIp) target.set("true-client-ip", trueClientIp);

  const clientIp = resolveForwardedClientIp(source);
  if (!clientIp) return;

  target.set("x-real-ip", clientIp);
  target.set("x-forwarded-for", clientIp);
}

File: apps/api/app/api/client_ip.py
from __future__ import annotations

from fastapi import Request

CLIENT_IP_HEADER_CANDIDATES = (
    "cf-connecting-ip",
    "true-client-ip",
    "x-forwarded-for",
    "x-real-ip",
)


def _normalize_ip_candidate(value: str | None) -> str | None:
    if not isinstance(value, str):
        return None
    candidate = value.split(",")[0].strip()
    if candidate == "":
        return None
    return candidate[:64]


def extract_request_client_ip(request: Request) -> str | None:
    for header_name in CLIENT_IP_HEADER_CANDIDATES:
        candidate = _normalize_ip_candidate(request.headers.get(header_name))
        if candidate:
            return candidate
    if request.client and request.client.host:
        return _normalize_ip_candidate(request.client.host)
    return None


def extract_request_client_ip_or_localhost(request: Request) -> str:
    return extract_request_client_ip(request) or "127.0.0.1"
